### PR TITLE
Sector and Industry Analysis Improvement

### PR DESCRIPTION
**Pull request by DidierRLopes**
This PR does the following:
- Improve parameters selection based on what has already been selected, e.g. if the sector is Financial Services, there's no point in showing as industry "Auto Manufacturers. Or if we choose the industry "Asset Management" we know that the sector is "Financial Services"
- Add this logic to the model module so that it speeds up the data retrieval because the loop to run is smaller
- Allows to study companies landscape in terms of companies per country using both sector or industry, and also companies per industry using sector
- Add multiple commands for additional financial stats like PEG, full time employees, market cap, beta, yield, dividend, ...
- Optimize code for these financial stats
- Fix some bugs

### CODE REVIEW

**Review by aia — APPROVED** [review]

**Comment by jmaslek** [comment]
So I just tried a random command to start and it resulted in an 8 minutes download.

Is it possible to have a "safety" somewhere when there is no sector/industry selected?


Screen Shot 2021-12-05 at 2 53 51 PM

**Comment by jmaslek** [comment]
```
(✨) (stocks)>(sia)> sector
show_options() got an unexpected keyword argument 'country' 

```

**Comment by piiq** [comment]
> Is it possible to have a "safety" somewhere when there is no sector/industry selected?

@jmaslek We have a progress bar, does that count?

**Comment by jmaslek** [comment]
> @jmaslek We have a progress bar, does that count?

lol.  I personally would make at least a Sector required for these.

**Comment by jmaslek** [comment]
My main comment is that there are just a lot of financial function options.  Not sure if better to have just one function with a variable flag (kind of like in the portfolio opt -- where user can select which yfinance info variable to use)

**Comment by DidierRLopes** [comment]
> So I just tried a random command to start and it resulted in an 8 minutes download.
> 
> Is it possible to have a "safety" somewhere when there is no sector/industry selected?
> 
> Screen Shot 2021-12-05 at 2 53 51 PM

Not really a safety because this is a valid choice. Best thing we can do is have a sector and industry per default? What you think?

**Comment by DidierRLopes** [comment]
> how_options(

Did you update `financedatabase`?

**Comment by DidierRLopes** [comment]
> > @jmaslek We have a progress bar, does that count?
> 
> lol. I personally would make at least a Sector required for these.

I'm not a fan of having a sector required because you may want to get all large caps of a country regardless of sector/industry and see which one has best PEG for instance... We ideally want to be as flexible as possible. But for people using this the first time (like you) they may use one of the financials without realising that it will take ages. Thus, I can select a Sector by default.

**Comment by piiq** [comment]
> Best thing we can do is have a sector and industry per default?

What sector/industry would you choose?

**Comment by piiq** [comment]
The only thing that annoys me is that after I download the data for 700+ tickers I can't export it to excel/csv and using the `--export` flag would require me to download the data all over again. Apart from that I think that it's ok

**Comment by jmaslek** [comment]
> Not really a safety because this is a valid choice. Best thing we can do is have a sector and industry per default? What you think?

Don't like.

**Comment by jmaslek** [comment]
> > how_options(
> 
> Did you update `financedatabase`?

oops

**Comment by DidierRLopes** [comment]
> The only thing that annoys me is that after I download the data for 700+ tickers I can't export it to excel/csv and using the `--export` flag would require me to download the data all over again. Apart from that I think that it's ok

Why can't you?

**Comment by DidierRLopes** [comment]
> > Not really a safety because this is a valid choice. Best thing we can do is have a sector and industry per default? What you think?
> 
> Don't like.

I tried it and think it is actually a good idea. @piiq what is your vote:
- Keep the same
- Add Sector/Industry analysis per default

Having a safety is not an option because this limits what the user can do. And we want to improve user experience, not limit their functionalities.

**Comment by piiq** [comment]
> Why can't you?

~~How? My assumption is that I call the command with `--export csv` arg to save the dataframe. And how to I export after I call the command? In my case I have called `roe` for all US large cap stocks, it downloaded data for 756, but displayed only 10. Is there a way for me to export a csv without re-running the `roe` that takes around 25 minutes (756 tickers * 2 seconds per ticker / 60 seconds)?~~ figured it out

> what is your vote

keep as it is now

> Having a safety is not an option because this limits what the user can do.

I think that we should think of safety from another angle. There is no harm if the user accidentally starts a 5 minute loop because if that's not the intended action the user can simply ctrl+c and restart the terminal. Harm comes if the data was inaccurate and the user lost money because of it.

p.s. 
1. selected the country (russia), switched the `exchange` flag and triggered `beta` and that resulted in
image
*edit whole writing:* there is a bunch of these
image
image

2. tried `roe` for all russian large cap and the chart is way off because the names are ridiculously long
image

**Comment by DidierRLopes** [comment]
> > Why can't you?
> 
> ~How? My assumption is that I call the command with `--export csv` arg to save the dataframe. And how to I export after I call the command? In my case I have called `roe` for all US large cap stocks, it downloaded data for 756, but displayed only 10. Is there a way for me to export a csv without re-running the `roe` that takes around 25 minutes (756 tickers * 2 seconds per ticker / 60 seconds)?~ figured it out _edit2:_ actually the question remains. if I call the same command with the --export flag it doesn't query the data anew, but only the top 10 are saved. is there some flag that i'm missing?
> 
> > what is your vote
> 
> keep as it is now
> 
> > Having a safety is not an option because this limits what the user can do.
> 
> I think that we should think of safety from another angle. There is no harm if the user accidentally starts a 5 minute loop because if that's not the intended action the user can simply ctrl+c and restart the terminal. Harm comes if the data was inaccurate and the user lost money because of it.
> 
> p.s.
> 
> 1. selected the country (russia), switched the `exchange` flag and triggered `beta` and that resulted in
>    image
>    _edit whole writing:_ there is a bunch of these
>    image
>    image
> 2. tried `roe` for all russian large cap and the chart is way off because the names are ridiculously long
>    image

I can fix the export, this was from my implementation.

Regarding the default, I think we should have them. Because our current defaults takes 5 seconds, so if someone tries this menu by trying one of the metrics they wait 5 seconds they see the output and then they can clear industry/sector and do wtv. Having to do Ctrl+C is not ideal, and 8min is a long ass time to wait.

Maybe the tickers you were looking for didn't have such data on YF? What were the parameters do you remember?

**Comment by piiq** [comment]
> Maybe the tickers you were looking for didn't have such data on YF? What were the parameters do you remember?

```
sia
country Russia
beta
etc...
```
Yeah, most likely they don't have this data in YF

> Regarding the default, I think we should have them.
Let's have 
```
│ Sector                         │ TRADE & SERVICES                           │
│ Industry                      │ RETAIL-COMPUTER & COMPUTER SOFTWARE STORES │
```
because GME? lol

**Comment by DidierRLopes** [comment]
> > Maybe the tickers you were looking for didn't have such data on YF? What were the parameters do you remember?
> 
> ```
> sia
> country Russia
> beta
> etc...
> ```
> 
> Yeah, most likely they don't have this data in YF
> 
> > Regarding the default, I think we should have them.
> 
> Let's have
> 
> ```
> │ Sector                         │ TRADE & SERVICES                           │
> │ Industry                      │ RETAIL-COMPUTER & COMPUTER SOFTWARE STORES │
> ```
> 
> because GME? lol

I dig that! Will do!

**Comment by DidierRLopes** [comment]
@piiq 

```
Industry          : Specialty Retail
Sector            : Consumer Cyclical
Country           : United States
Market Cap        : Large
Exclude Exchanges : True
```

this is GME one

**Comment by piiq** [comment]
@DidierRLopes maybe we disable commands with empty sector/industry but add some sort of an `all` flag?

**Comment by DidierRLopes** [comment]
> @DidierRLopes maybe we disable commands with empty sector/industry but add some sort of an `all` flag?

Nahh, don't like that addition of complexity. Try out what we have and give feedback tomorrow on it please!

**Comment by piiq** [comment]
@DidierRLopes I've allowed myself to commit a minor change to the way the chart title text is set in order for the cart title to fit the screen space. Please pull the changes

In addition, I think that the `--export` is not working as expected after the latest updates. If I run `metric -m eg --export csv` on the default industry/sector I get a csv with 9 rows, but if I toggle the exchange flag and run the same command I expect to see 35 rows in the csv, but still get only 9.
I've tried to run `metric -m eg -l 100 --export csv` but the result was the same (9 rows)
Let me know if I'm missing anything

**Comment by DidierRLopes** [comment]
Is it possible that there are only 9 companies? @piiq 

Can you show me what did you use? when I tested the export was working fine but my testing wasn't extensive.

Screenshot 2021-12-06 at 10 44 09

I improved even further the title, have a peek. Thanks!

**Comment by piiq** [comment]
> Can you show me what did you use? when I tested the export was working fine but my testing wasn't extensive.

```
stocks
sia
metric eg                                 <- samples 9 companies
exchange
metric eg                                 <- samples 35 companies
metric eg --export csv           <- saves 9 companies
```

Re: charts can I add more stuff? I want to split long company names to multiple lines

**Comment by DidierRLopes** [comment]
> > Can you show me what did you use? when I tested the export was working fine but my testing wasn't extensive.
> 
> ```
> stocks
> sia
> metric eg                                 <- samples 9 companies
> exchange
> metric eg                                 <- samples 35 companies
> metric eg --export csv           <- saves 9 companies
> ```
> 
> Re: charts can I add more stuff? I want to split long company names to multiple lines

Let me fix the export and then you can improve charting!

**Comment by DidierRLopes** [comment]
Screenshot 2021-12-06 at 11 15 57

Although it iterates through 35 companies only 9 of them actually have the "eg" metric, hence why you are only exporting and seeing 9. The other hint is that we aren't receiving the "limiting of 10 companies"

**Comment by DidierRLopes** [comment]
@piiq green light for improving chart!

**Comment by jmaslek** [comment]
Some metrics get a numpy error:

```

(✨) (stocks)>(sia)> metric mc
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 9/9 [00:06<00:00,  1.32it/s]
'numpy.ndarray' object has no attribute 'div' 

(✨) (stocks)>(sia)> metric td
100%|████████████████████████████████████████████████████████████████████████████████████████████| 9/9 [00:06<00:00,  1.37it/s]
'numpy.ndarray' object has no attribute 'div' 
```

metric eg gave me no error (this is all with default set)

**Comment by piiq** [comment]
> Some metrics get a numpy error:

I was just debugging this. It happens here
image

**Comment by jmaslek** [comment]
> Because .div comes from a df, and here we have an ndarray

Yup.  Can be replaced with .divide() -- or just do /= 1_000

**Comment by piiq** [comment]
> Can be replaced with .divide()

replaced this with `company_metric = np.divide(company_metric, 1_000)`, and it checks out. Will commit a bit later

**Comment by DidierRLopes** [comment]
I'm retard ahah changed the list to numpy and though .div would work. And metrics I tested didn't need it hence why I didn't see it.

Will wait for your commit @piiq, let me know if you want me to do it

**Comment by piiq** [comment]
@DidierRLopes I've made a couple of changes here. Most important one is handling of missing data. Now it shouldn't break the loop execution if one of the many companies in the query has a missing metric or some other missing data.

And I've split long company names into 2 lines on the chart. This applies to names that have more that 6 words with more than 40 characters

**Review by piiq — APPROVED** [review]
A very good menu we have here. Need to add building sector maps in the future

File: gamestonk_terminal/stocks/sector_industry_analysis/financedatabase_view.py [hunk 1]
```diff
@@ -96,13 +72,10 @@ def display_bars_financials(
 
     metric_data = {}
     for symbol in list(stocks_data.keys()):
-        if (
-            "financialData" in stocks_data[symbol]
-            and "quoteType" in stocks_data[symbol]
-        ):
-            metric = stocks_data[symbol]["financialData"][finance_metric]
+        if finance_key in stocks_data[symbol] and "quoteType" in stocks_data[symbol]:
+            metric = stocks_data[symbol][finance_key][finance_metric]
```
[the remark this answers is not in the document]
Comment (piiq) [reply gamestonk_terminal/stocks/sector_industry_analysis/financedatabase_view.py]: this is fixed

[the remark this answers is not in the document]
Comment (DidierRLopes) [reply gamestonk_terminal/stocks/sector_industry_analysis/financedatabase_view.py]: I was pretty sure I had fixed this one!

File: gamestonk_terminal/stocks/sector_industry_analysis/financedatabase_view.py [hunk 2]
```diff
@@ -73,8 +73,13 @@ def display_bars_financials(
     metric_data = {}
     for symbol in list(stocks_data.keys()):
         if finance_key in stocks_data[symbol] and "quoteType" in stocks_data[symbol]:
-            metric = stocks_data[symbol][finance_key][finance_metric]
             stock_name = stocks_data[symbol]["quoteType"]["longName"]
+            metric = (
+                stocks_data[symbol][finance_key][finance_metric]
+                if stocks_data[symbol][finance_key] is not None
+                and finance_metric in stocks_data[symbol][finance_key]
+                else None
```
Comment (DidierRLopes) [inline gamestonk_terminal/stocks/sector_industry_analysis/financedatabase_view.py:81]: gotcha! Good one!

**Pull request merged**